micropython/bluetooth/aioble/aioble/central.py

```diff
@@ -187,10 +188,19 @@ def _decode_field(self, *adv_type):
                     yield payload[i + 2 : i + payload[i] + 1]
                 i += 1 + payload[i]
 
-    # Returns the value of the advertised name, otherwise empty string.
+    # Returns the value of the complete (or shortened) advertised name, otherwise empty string.
     def name(self):
         for n in self._decode_field(_ADV_TYPE_NAME):
-            return str(n, "utf-8") if n else ""
+            if n:
```

Comment:
Thanks @FlantasticDan, I've never seen a device using the short name which is why I never added it, but this looks useful.

The `_decode_field` function allows multiple field IDs to be specified, but they are returned in payload order. I don't see any reason not to re-order the loops in `_decode_field` though to make them in `*arg` order instead though (this wouldn't affect the `services` function because it's returning all services anyway). So that would make this a much smaller change to support short name.

Do you think it's likely that a device would include both short and long names though? Perhaps short in the adv payload and long in the scan response? This would mean that a device would return the short name in the first iteration of the scanner loop, and the long name subsequently. So should we make this a separate function?

FWIW, one thing I've been considering is not returning results for active scans for devices that support scan responses until the scan response arrives (this is what Android and iOS do). This "feature" of aioble to return the partial result is a source of much confusion. My intention had been that an app shouldn't care -- it just keeps scanning until it matches what it needs.

Reply:
@jimmo appreciate the review!  I didn't realize `_decode_field` could take multiple field IDs.  That given with the below excerpt of the spec leads me to believe we should attempt to decode both long and short name and just return whatever comes back first.

> The Local Name data type shall be the same as, or a shortened version of, the local name assigned to the device. The Local Name data type value indicates if the name is complete or shortened. If the name is shortened, the complete name can be read using the remote name request procedure over BR/EDR or by reading the device name characteristic after the connection has been established using GATT.
> 
> A shortened name shall only contain contiguous characters from the beginning of the full name. For example, if the device name is ‘BT_Device_Name’ then the shortened name could be ‘BT_Device’ or ‘BT_Dev’.
>
> __*from the* Supplement to the Bluetooth Core Specification v11, Part A, Section 1.2.1__

I think the spec implies that devices will only send either the complete **or** short name because it explicitly describes the procedure to get a long name when given a short name.  Combine that with the fact that the short name must be contiguous characters from the complete name I think even if in the edge case where the user was returned a short name when a long name was available, it would still be clear which device they were communicating with and thus a non-issue.  What do you think?

I'm happy to modify this PR in this direction.

---

> FWIW, one thing I've been considering is not returning results for active scans for devices that support scan responses until the scan response arrives (this is what Android and iOS do). This "feature" of aioble to return the partial result is a source of much confusion. My intention had been that an app shouldn't care -- it just keeps scanning until it matches what it needs.

Correct me if I'm wrong but I think this explains why an active scan will return the same device (as identified by it's address) and sometimes be able to populate the name field and other times return an empty string...  I've been using the following excerpt to I suspect accommodate the "feature" you described:
```python
devices = []
addrs = set()
async with aioble.scan(duration_ms=5000, interval_us=30000, window_us=30000, active=True) as scanner:
    async for result in scanner:
        if result.device.addr_type == 0 and result.name() != '' and result.device.addr_hex() not in addrs:
            devices.append((result.name().rstrip('\x00'), result.device.addr_hex()))
            addrs.add(result.device.addr_hex())
print(devices)
```

Not that you asked but if you are considering changing this behavior perhaps a non breaking way to do it would be to add a kwarg like `include_partial=True` so that users that want the more iOS/Android-y experience can opt in to it.

Reply:
> I'm happy to modify this PR in this direction.

That would be good thanks! (So I think that means this PR just changes that one line to `for n in self._decode_field(_ADV_TYPE_NAME, _ADV_TYPE_SHORT_NAME):` ?

> Correct me if I'm wrong but I think this explains why an active scan will return the same device (as identified by it's address) and sometimes be able to populate the name field and other times return an empty string...

Yep, that's exactly right. My theory was that if you need the full information (only available once the scan result arrives) then you would just ignore the first result. And if the first result was enough, then that's great you can connect immediately.

> Not that you asked but if you are considering changing this behavior perhaps a non breaking way to do it would be to add a kwarg like include_partial=True so that users that want the more iOS/Android-y experience can opt in to it.

Yes, perhaps. (But maybe I didn't make it clear -- Android and iOS suppress the results _until_ the scan response arrives. So in order to make it backwards-compatible, the Android/iOS behavior would be opt-in).